staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+	// short-circuit on exempt runtimeclass
+	if a.exemptRuntimeClass(podSpec.RuntimeClassName) {
+		return allowedResponse()
+	}
+
+	namespace, err := a.NamespaceGetter.GetNamespace(ctx, podMetadata.Namespace)
+	if err != nil {
+		klog.ErrorS(err, "failed to fetch pod namespace", "namespace", podMetadata.Namespace)
+		return internalErrorResponse(fmt.Sprintf("failed to lookup namespace %s", podMetadata.Namespace))
+	}
+
+	auditAnnotations := map[string]string{}
+	policy, err := a.PolicyToEvaluate(namespace.Labels)
+	if err != nil {
+		klog.V(2).InfoS("failed to parse PodSecurity namespace labels", "err", err)
+		auditAnnotations["error"] = fmt.Sprintf("Failed to parse policy: %v", err)
+	}
+
+	response := allowedResponse()
+	if enforce {
+		if result, err := a.evaluateLevelAndVersion(podMetadata, podSpec, policy.Enforce); err != nil {
+			klog.ErrorS(err, "Enforce evaluation failed", "namespace", namespace.Name, "pod", podMetadata.Name, "profile", policy.Enforce.String())
+			response = internalErrorResponse(err.Error())
+		} else if !result.Allowed {
+			response = forbiddenResponse(result.ForbiddenDetail())
+		}
+	}
+
+	// TODO: reuse previous evaluation if audit level+version is the same as enforce level+version
+	if result, err := a.evaluateLevelAndVersion(podMetadata, podSpec, policy.Audit); err != nil {
+		klog.ErrorS(err, "Audit evaluation failed", "namespace", namespace.Name, "pod", podMetadata.Name, "profile", policy.Audit.String())
+		auditAnnotations["audit-error"] = err.Error()
+	} else if !result.Allowed {
+		auditAnnotations["audit"] = result.ForbiddenDetail()
+	}
+
+	// TODO: reuse previous evaluation if warn level+version is the same as audit or enforce level+version
+	if result, err := a.evaluateLevelAndVersion(podMetadata, podSpec, policy.Warn); err != nil {
+		klog.ErrorS(err, "Warn evaluation failed", "namespace", namespace.Name, "pod", podMetadata.Name, "profile", policy.Warn.String())
+	} else if !result.Allowed {
+		// TODO: Craft a better user-facing warning message
+		response.Warnings = append(response.Warnings, fmt.Sprintf("Pod violates PodSecurity profile %s: %s", policy.Warn.String(), result.ForbiddenDetail()))
+	}
+
+	response.AuditAnnotations = auditAnnotations
+	return response
+}
+
+func (a *Admission) evaluateLevelAndVersion(podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec, lv api.LevelVersion) (policy.AggregateCheckResult, error) {
+	// TODO: optimize so we don't have to construct/return a list here three times per request
+	checks, err := a.Registry.ChecksForLevelAndVersion(lv.Level, lv.Version)
+	if err != nil {
+		return policy.AggregateCheckResult{}, fmt.Errorf("failed to get checks: %v", err)
+	}
+
+	return policy.AggregateCheckPod(checks, podMetadata, podSpec), nil
+}
+
+func (a *Admission) CheckPodsInNamespace(ctx context.Context, namespace string, enforce api.LevelVersion) []string {
```

Comment:
note to self, you didn't review this because you're trying to find the check registry before 10am